### PR TITLE
[FVM] signature verification refactoring

### PR DESCRIPTION
**Pull request by tarakby**
- remove the user-domain tag completely from the repo, as this is not a protocol constant anymore. 
- remove not needed `SignatureVerifier` interface to verify signatures. 
- separate the verification functions of runtime and transaction signatures that support different algorithms for a better clarity and to avoid future errors.
- add extra checks for transaction verification inputs (although they are supposed to be valid as they are read from the state).
- make the public function `SignWithTag` unable to sign other that the transaction tag for safety. 
- adjust existing tests, add new tests
- avoid using `fmt.Sprintf` for key deduplication (performance)

### CODE REVIEW

File: fvm/crypto/crypto.go [hunk 1]
```diff
@@ -160,71 +164,83 @@ func VerifySignatureFromRuntime(
...
+	valid, err := pk.Verify(signature, message, hasher)
 	if err != nil {
 		// All inputs are guaranteed to be valid at this stage.
 		// The check for crypto.InvalidInputs is only a sanity check
 		if crypto.IsInvalidInputsError(err) {
 			return false, err
 		}
-		panic(fmt.Errorf("verify signature failed with unexpected error %w", err))
+		// TODO: check if we should panic or bubble the error up
+		panic(fmt.Errorf("verify transaction signature failed with unexpected error %w", err))
```
Comment (tarakby) [inline fvm/crypto/crypto.go:243]: @ramtinms, wondering if we should panic or return an error here. This is a case of unexpected error (error in allocating memory for instance..)

File: fvm/crypto/crypto.go [hunk 2]
```diff
@@ -160,71 +164,83 @@ func VerifySignatureFromRuntime(
...
+	if pk.Algorithm() != crypto.ECDSAP256 && pk.Algorithm() != crypto.ECDSASecp256k1 {
+		// TODO: check if we should panic
+		// This case only happens in production if there is a bug
+		return false, errors.NewValueErrorf(pk.Algorithm().String(), "is not supported in transactions")
+	}
+	// hashing compatibility
+	if hashAlgo != hash.SHA2_256 && hashAlgo != hash.SHA3_256 {
+		// TODO: check if we should panic
+		// This case only happens in production if there is a bug
+		return false, errors.NewValueErrorf(hashAlgo.String(), "is not supported in transactions")
```
Comment (tarakby) [inline fvm/crypto/crypto.go:227]: @ramtinms : If this error happens then there is a bug. Should we panic?

Comment (janezpodhostnik) [reply fvm/crypto/crypto.go:227]: I think a FVM `errors.Failure` should be returned instead.

File: fvm/crypto/hash.go [hunk 3]
```diff
@@ -17,38 +20,36 @@ import (
...
-	case hash.SHA2_384:
-		hasher = hash.NewSHA2_384()
-	case hash.SHA3_384:
-		hasher = hash.NewSHA3_384()
-	case hash.Keccak_256:
-		hasher = hash.NewKeccak_256()
-	default:
-		return nil, errors.New("hashing algorithm is not a supported for prefixed algorithm")
+	hasherCreator := hasherCreators[algo]
+	if hasherCreator == nil {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply fvm/crypto/hash.go:51]: Yep, good suggestion 👌🏼

File: access/validator.go [hunk 4]
```diff
@@ -267,13 +267,16 @@ func (v *TransactionValidator) checkAddresses(tx *flow.TransactionBody) error {
...
+	}
+	observedSigs := make(map[uniqueKey]bool)
 	for _, sig := range append(tx.PayloadSignatures, tx.EnvelopeSignatures...) {
-		keyStr := sig.UniqueKeyString()
-		if observedSigs[keyStr] {
+		if observedSigs[uniqueKey{sig.Address, sig.KeyIndex}] {
 			return DuplicatedSignatureError{Address: sig.Address, KeyIndex: sig.KeyIndex}
 		}
-		observedSigs[keyStr] = true
+		observedSigs[uniqueKey{sig.Address, sig.KeyIndex}] = true
```
Comment (ramtinms) [inline access/validator.go:279]: I assume this change is mostly for performance reasons, could you please check we have tests for this?

File: fvm/context.go [hunk 5]
```diff
@@ -32,7 +31,6 @@ type Context struct {
 	ServiceEventCollectionEnabled bool
 	AccountFreezeAvailable        bool
 	ExtensiveTracing              bool
-	SignatureVerifier             crypto.SignatureVerifier
```
Comment (ramtinms) [inline fvm/context.go:35]: this change might need some changes on the emulator and other places that has their own context and environments.

File: model/flow/constants.go [hunk 6]
```diff
@@ -42,22 +42,25 @@ const DefaultMaxAddressIndex = 20_000_000
...
+//
+// Signatures on Flow that needs to be scoped to a certain domain need to
+// have the same length in order to avoid tag collision issues, when prefixing the
+// message to sign.
 const DomainTagLength = 32
 
 const TransactionTagString = "FLOW-V0.0-transaction"
 
-const UserTagString = "FLOW-V0.0-user"
-
```
Comment (ramtinms) [inline model/flow/constants.go:50]: this might be in use by go-sdk or others ?

File: fvm/crypto/crypto.go (hunk 1 above)
Comment (ramtinms) [reply fvm/crypto/crypto.go:245]: we usually panic if there is a small chance of non-deterministic reasons that caused this error.

File: fvm/crypto/crypto.go [hunk 7]
```diff
@@ -160,71 +164,85 @@ func VerifySignatureFromRuntime(
...
+		return false, errors.NewUnknownFailure(fmt.Errorf(
+			pk.Algorithm().String(), "is not supported in transactions"))
+	}
+	// hashing compatibility
+	if hashAlgo != hash.SHA2_256 && hashAlgo != hash.SHA3_256 {
+		// TODO: check if we should panic
+		// This case only happens in production if there is a bug
+		return false, errors.NewUnknownFailure(fmt.Errorf(
+			hashAlgo.String(), "is not supported in transactions"))
+	}
```
Comment (ramtinms) [inline fvm/crypto/crypto.go:230]: not sure about this, you mean if there is some problematic old keys we would stop the network? I think we should just error and add a watch on the error type.

File: fvm/crypto/crypto.go [hunk 8]
```diff
@@ -146,11 +154,7 @@ func VerifySignatureFromRuntime(
 			return false, errors.NewValueErrorf(sigAlgo.String(), "cannot use hashing algorithm type %s with signature signature algorithm type %s",
 				hashAlgo, sigAlgo)
 		}
-
-		// tag length compatibility
-		if len(tag) > flow.DomainTagLength {
-			return false, errors.NewValueErrorf(tag, "tag length (%d) is larger than max length allowed (%d bytes).", len(tag), flow.DomainTagLength)
-		}
+		// tag constraints are checked when initializing a prefix-hasher
```
Comment (ramtinms) [inline fvm/crypto/crypto.go:157]: you mean we check the constraints inside NewPrefixedHashing ?

File: fvm/crypto/crypto.go [hunk 9]
```diff
@@ -160,71 +164,85 @@ func VerifySignatureFromRuntime(
...
-	publicKey crypto.PublicKey,
+	pk crypto.PublicKey,
 	hashAlgo hash.HashingAlgorithm,
 ) (bool, error) {
 
-	var hasher hash.Hasher
+	// check ECDSA compatibilites
+	if pk.Algorithm() != crypto.ECDSAP256 && pk.Algorithm() != crypto.ECDSASecp256k1 {
+		// TODO: check if we should panic
+		// This case only happens in production if there is a bug
```
Comment (ramtinms) [inline fvm/crypto/crypto.go:220]: we should probably run some analysis on the mainnet data to verify all data are okey

**Review by ramtinms — APPROVED** [review]
Looks good to me. thanks for clearing things up.

File: model/flow/transaction.go [hunk 10]
```diff
@@ -273,16 +273,18 @@ func (tb *TransactionBody) SignEnvelope(
...
 
-// SignMessageWithTag signs the message (tag + message) with the specified account key.
+// SignMessageWithTag signs the data (transaction_tag + message) with the specified private key
+// and hasher.
 //
-// This function returns an error if the signature cannot be generated.
+// This function returns an error if:
+//  - crypto.InvalidInputsError if the private key cannot sign with the given hasher
+//  - other error if an unexpected error occurs
 func (tb *TransactionBody) SignMessageWithTag(
```
Comment (janezpodhostnik) [inline model/flow/transaction.go:282]: Does it make sense to rename this to `SignTransactionMessage`?

File: model/flow/constants.go (hunk 6 above)
Comment (janezpodhostnik) [reply model/flow/constants.go:50]: It is in use, this is a breaking change for the `flow-go-sdk/examples`. We should keep that in mind.

File: fvm/transactionVerifier.go [hunk 11]
```diff
@@ -42,6 +33,14 @@ func (v *TransactionSignatureVerifier) Process(
 	return v.verifyTransactionSignatures(proc, *ctx, sth)
 }
 
+func newInvalidEnvelopeSignatureError(txSig flow.TransactionSignature, err error) error {
```
Comment (janezpodhostnik) [inline fvm/transactionVerifier.go:36]: Nice, I like this approach!

File: fvm/crypto/hash.go [hunk 12]
```diff
@@ -17,38 +20,36 @@ import (
...
 
 // paddedDomainTag converts a string into a padded byte array.
-func paddedDomainTag(s string) ([flow.DomainTagLength]byte, error) {
-	var tag [flow.DomainTagLength]byte
-	if len(s) > flow.DomainTagLength {
-		return tag, fmt.Errorf("domain tag %s cannot be longer than %d characters", s, flow.DomainTagLength)
+func paddedDomainTag(s string) ([tagLength]byte, error) {
+	var tag [tagLength]byte
+	if len(s) > tagLength {
+		return tag, fmt.Errorf("domain tag cannot be longer than %d characters, got %d", tagLength, len(s))
```
Comment (janezpodhostnik) [inline fvm/crypto/hash.go:30]: The error no longer includes the tag text. Is this ok?

**Review by janezpodhostnik — APPROVED** [review]
Some minor comments.

File: access/validator.go (hunk 4 above)
Comment (tarakby) [reply access/validator.go:279]: Yes, it is for performance. Here are the tests https://github.com/onflow/flow-go/blob/master/fvm/transactionVerifier_test.go#L43-L68

File: fvm/context.go (hunk 5 above)
Comment (tarakby) [reply fvm/context.go:35]: I already checked the emulator doesn't use the interface. If it does, I'll also update the emulator. 
Is there another tool that you think uses its own context?

File: model/flow/constants.go (hunk 6 above)
Comment (tarakby) [reply model/flow/constants.go:50]: I'll update the sdk too after merging this PR: https://github.com/onflow/flow-go-sdk/issues/273

File: fvm/crypto/crypto.go (hunk 7 above)
Comment (tarakby) [reply fvm/crypto/crypto.go:230]: I tagged you about this [here](https://github.com/onflow/flow-go/pull/2171#discussion_r832795320). I suggested to panic, Janez suggested the Failure error. As mentioned in the comment, that case can only happen if there is a bug in the code. These keys are all account keys, they are supposed to be checked and added previously, and they are not supposed to be problematic. 

What do you suggest to use here?

File: fvm/crypto/crypto.go (hunk 8 above)
Comment (tarakby) [reply fvm/crypto/crypto.go:157]: Correct

File: fvm/crypto/crypto.go (hunk 9 above)
Comment (tarakby) [reply fvm/crypto/crypto.go:220]: These are all keys already checked and added using `AddAccountKey`, so this check is only a sanity check, and if it fails then we have am internal bug somewhere.

This is the same as above. I suggested to panic and Janez suggest `Failure` error. What do you suggest?

File: model/flow/transaction.go (hunk 10 above)
Comment (tarakby) [reply model/flow/transaction.go:282]: How about just `Sign`, so it would be a method `TransactionBody.Sign` ?

File: fvm/crypto/hash.go (hunk 12 above)
Comment (tarakby) [reply fvm/crypto/hash.go:30]: I will add it back, that's better.

**Review by janezpodhostnik — APPROVED** [review]

File: model/flow/transaction.go (hunk 10 above)
Comment (tarakby) [reply model/flow/transaction.go]: https://github.com/onflow/flow-go/pull/2171/commits/dcd8a01ce5f9031960837913cf9fccb0a4e0f859

**Review by ramtinms — APPROVED** [review]
Looks good to me

**Comment by tarakby** [comment]
bors merge

**Pull request merged**